pkg/minishift/update/update.go

```diff
@@ -0,0 +1,305 @@
+/*
+Copyright (C) 2016 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package update
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/google/go-github/github"
+	update "github.com/inconshreveable/go-update"
+	"github.com/minishift/minishift/pkg/util/archive"
+	pb "gopkg.in/cheggaaa/pb.v1"
+
+	githubutils "github.com/minishift/minishift/pkg/util/github"
+	"github.com/minishift/minishift/pkg/version"
+)
+
+const (
+	timeLayout  = time.RFC1123
+	githubOwner = "minishift"
+	githubRepo  = "minishift"
+)
+
+// CurrentVersion returns the current version of minishift binary installed on the system
+func CurrentVersion() (semver.Version, error) {
+	localVersion, err := version.GetSemverVersion()
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return localVersion, nil
+}
+
+// LatestVersion returns the latest version of minishift binary available from upstream
+func LatestVersion() (semver.Version, error) {
+	latestVersion, err := getLatestVersionFromGitHub(githubOwner, githubRepo)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return latestVersion, nil
+}
+
+// IsNewerVersion compares the local and latest versions and returns a boolean
+func IsNewerVersion(localVersion, latestVersion semver.Version) bool {
+	if localVersion.Compare(latestVersion) < 0 {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Update handles the update process by calling functions that download, verify, extract and replace the binary.
+// It returns an error if any of these functions fails at any point.
+func Update(output io.Writer, latestVersion semver.Version) error {
+	var extName string
+
+	if runtime.GOOS == "windows" {
+		extName = "zip"
+	} else {
+		extName = "tgz"
+	}
+
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", latestVersion, runtime.GOOS, runtime.GOARCH, extName)
+	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+
+	url := fmt.Sprintf(downloadLinkFormat, latestVersion, archiveName)
+	downloadedArchivePath, err := downloadAndVerifyArchive(url)
+	if err != nil {
+		return err
+	}
+
+	// Get the directory containing saved archive
+	dir := strings.Split(downloadedArchivePath, "minishift")[0]
```

Comment:
This is prone to break and makes assumptions about the code in _downloadAndVerifyArchive_. Let's create the tmp directory here and pass it to _downloadAndVerifyArchive_. This was we don't have to make assumptions and know exactly we directory we need to delete on defer.